### ISSUE TITLE
Should gophernicus serve a document root specified with a relative path?

### ISSUE THREAD

**Issue by augfab** [gophernicus/gophernicus#71]
Hi,
Because my memory is bad, I keep running into the same problem when I test `gophernicus` directly in the terminal: it `chdir(2)`'s to the document root directory, then tries to find a gophermap with name `<document_root>/gophermap`.
If `<document_root>` is a relative path (relative to the directory from which `gophernicus` is invoked), then `<document_root>` will be in the relative path “twice”.

Incriminated code is `src/menu.c` line 478:
```c
	/* Check for a gophermap */
	snprintf(pathname, sizeof(pathname), "%s/%s",
		st->req_realpath, st->map_file);
```

Steps to reproduce (with `a6df879d6186ba78f7079cc54dc545b3635e5d96`):
```
$ make clean && make
$ (cd ./.travis && ln -s test.gophermap gophermap)
$ printf '\n' >input
$ ./src/gophernicus -r ./.travis -nm -nf -d < input
i[/]	TITLE	null.host	1
i		null.host	1
.
```

What I expected to see:
```
iA		null.host	1
iLine		null.host	1
iRight		null.host	1
iHere		null.host	1
iFor		null.host	1
iTests		null.host	1
0file	file	localhost	70
.
```

When `gophernicus` is invoked with `-r .`, it works as expected.
```
$ cd ./.travis
$ ../src/gophernicus -r . -nm -d < ../input
```

So what should `gophernicus` do?
* `chdir(2)` to `<document_root>` and look for the gophermap relative to `<selector>` (sanitized to remove `..` and such)?
* Stay in the initial working directory, and prefix the selector with the `<document_root>` (absolute or relative, it doesn't matter).

I prefer the first option.

What do you think?

**Comment by augfab** [comment]
Two other possibilities would be to:
* Use the absolute path of the `<document_root>`, unsing `readlink(2)` for instance.
* Refuse to serve document roots specified with a relative path.

**Comment by fosslinux** [comment]
I also prefer the first option. `chdir(2)` is the best way forward here. I am not sure why you think it needs to be sanatized? The `-r` argument should always stay the same between each `gophernicus` call.

I'm not completely sure what this bug is... it seems `-r` is broken? for relative paths. I need to investigate.

**Comment by augfab** [comment]
I wrote some changes ([`b809972`](https://github.com/augfab/gophernicus/commit/b8099726c58112d42c76cd8243795556fa4ec6fe) and parents) which appear to fix the problem.
I tested it with a document root specified with a relative path (`-r ./test` and `-r ../test`) and absolute path, serving text files and an executable gophermap, on Ubuntu 19.10.
However, these little tests are a long shot from covering all the code I changed.

I'll test some more scenarios, as I'm not fully confident this doesn't break other stuff.

I think in the future we should work on extending the test suite, and maybe add a `check` target to the Makefile to make it easier to run.

Also maybe we should add automatic code coverage collection?
